### ISSUE TITLE
How do I unsplit panes?

### ISSUE THREAD

**Issue by jbarnette** [atom/atom#380]
Splitting panes is easy: I got `C-|` and `C--` into my fingers quickly. What's hard for me is unsplits. Currently it looks like I have to `M-w` until I've closed every tab in the pane, at which point it goes away.

The problem for me is that I don't use tabs, and I don't think of files being open in a pane. I think of them as being open in _Atom_ and displayed in a pane when I ask for 'em.

When I think about unsplitting I don't connect it to the idea of closing files at all. I'd love it if unsplitting was a pane-level concern like splitting is.

**Comment by quasipedia** [comment]
Is still a pending issue 3 years later or the functionality has been implemented and this issue has remained open by mistake?  If it is the latter... how do I merge to panes?  I could not find the command in the command palette... :disappointed:

**Comment by nathansobo** [comment]
It's not implemented yet, but could be done pretty easily in a package until it bubbles to top priority in another 3 years :trollface:.

**Comment by quasipedia** [comment]
@nathansobo - I'm pretty new to writing atom packages (I published [my first one](https://atom.io/packages/geiger) last week).

If you can give me some general pointer (like the part of the application state to access, or what API, or any valuable link) I'd be happy to have a crack at it. :)

**Comment by jbarnette** [comment]
> another 3 years

@nathansobo 😀 Meet you here in 2019.

**Comment by wfarr** [comment]
> @nathansobo 😀 Meet you here in 2019.

I'll bring the 🍻

**Comment by nathansobo** [comment]
@quasipedia Take a look at the `Pane` class in Atom... you can read the API docs, but I'd also suggest looking at the source. You will need to use some private APIs to achieve this involving a class called `PaneAxis`, but this area of Atom is quite stable.

Basically, what you'd need to do to unsplit a pane is access the pane's parent. If the pane's parent is a `PaneAxis` rather than the root `PaneContainer` object, you'll want to grab the pane's closest sibling and move all the pane's items to it, then destroy the pane. I'd suggest trying it out in a package and getting the UX feeling good. At that point maybe you could submit a PR to core. But the package means you can start using the functionality right away without getting blocked on code review.

**Comment by quasipedia** [comment]
Thank you @nathansobo - I will have a crack at it once back from holidays. :)

**Comment by darwayne** [comment]
@jbarnette I just switched to atom from sublime .... based on @nathansobo suggestion you can do the following to get the same unsplit behavior as sublime:

init.coffee:
````
atom.commands.add 'atom-text-editor', 'pane:merge', ->
  panes     = atom.workspace.getPanes()
  firstPane = panes[0]
  for pane in panes
    if pane != firstPane
      for item in pane.getItems()
        isActive = pane.getActiveItem() == item && pane.isActive
        pane.removeItem(item)
        firstPane.addItem(item, {moved: true})
        if isActive
          firstPane.setActiveItem(item)

      pane.destroy()
````

keymap.cson:
````
'atom-text-editor':
  'cmd-alt-2': 'pane:split-right'
  'cmd-alt-1': 'pane:merge'
```

**Comment by quasipedia** [comment]
BTW.  I did have a crack at creating the package, but the behaviour was buggy and I did not get around to find the time to investigate more.  If anything is interested in trying this out / do a pull request, the code for the package is [here](https://github.com/quasipedia/unsplit-panes).

**Comment by mrchess** [comment]
@darwayne Wow great. I was looking exactly for this. Thanks. 

Is it working for you though? It seems to work well for me, but `pane:merge` shortcut is not working... hmm...

**Comment by mrchess** [comment]
I figured it out. I think the API slightly changed? Correct version...

```coffeescript
atom.commands.add 'atom-text-editor', 'pane:merge': -> # Changed this to object
  panes     = atom.workspace.getPanes()
  firstPane = panes[0]
  for pane in panes
    if pane != firstPane
      for item in pane.getItems()
        isActive = pane.getActiveItem() == item && pane.isActive()
        pane.removeItem(item)
        firstPane.addItem(item, {moved: true})
        if isActive
          firstPane.setActiveItem(item)

      pane.destroy()
```

**Comment by Abhi2424shek** [comment]
@mrchess updates pls

**Comment by rodriguezmanu** [comment]
+1

**Comment by JerryGoyal** [comment]
Ok, there's a package which does the job for combining tabs: [https://github.com/vidius/atom-pain-split](pane split) use `ctrl-k m` shortcut to combine panes.

**Comment by rbalicki2** [comment]
This closes the tree view as well. If you add this to your `init.js` (or use a coffee script converter to add it to your `init.coffee` file):

```
atom.commands.add('atom-text-editor', {
  'pane:merge': () => {
    const panes = atom.workspace.getPanes();
    const hasActiveItem = pane => pane.activeItem;
    const isTreePane = pane => (hasActiveItem(pane) && pane.activeItem.element.className.includes('tree-view'));
    const firstNonTreePane = panes.find(pane => !isTreePane(pane));

    panes.forEach(pane => {
      if (pane !== firstNonTreePane && hasActiveItem(pane) && !isTreePane(pane)) {
        pane.getItems().forEach(item => {
          const isActive = pane.getActiveItem() === item && pane.isActive();
          pane.removeItem(item);
          firstNonTreePane.addItem(item, { moved: true });
          if (isActive) {
            firstNonTreePane.setActiveItem(item);
          }
        });

        pane.destroy();
      }
    });
  },
});

```

It preserves your tree view pane

**Comment by lemongrassnginger** [comment]
It sounds like actually what we need to do is find out how to grab all panes that aren't in docks or panels. Other packages [such as this one](https://atom.io/packages/document-outline) can use that space to add tools etc which I guess we're not interested in merging when we talk about merging panes. Just ignoring the tree-view would miss other packages that use the docks. 

I don't properly understand the distinction between panes, panels and docks yet.

**Comment by lemongrassnginger** [comment]
so atom workspace provides WorkspaceCenter, a pane container that doesn't seem to include the docks/surrounding panels (still not sure whether those are different things yet). 

So this seems to work for me (in my init script)

```
# Merge panes
atom.commands.add 'atom-workspace', 'init:merge-panes', ->
  aPanes = atom.workspace.getCenter().getPanes()
  oFirstPane = aPanes.shift()
  for oPane in aPanes
    for oItem in oPane.getItems()
      oPane.moveItemToPane oItem, oFirstPane
```

**Comment by emmanuelito** [comment]
@lemongrassnginger thank you, works for me (binded to ctrl-x 0, as expected by an Emacs's guy).

**Comment by nwaywood** [comment]
I am trying to extend @lemongrassnginger example so that it excludes duplicates when merging. However, the code below doesn't work, anyone know what I'm doing wrong?

```
atom.commands.add 'atom-workspace', 'custom:merge-panes', ->
  panes = atom.workspace.getCenter().getPanes()
  firstPane = panes.shift()
  # loop through all panes except for the first pane
  for pane in panes
    for item in pane.getItems()
      # if item is already in first pane, delete it, otherwise move it to first pane
      for firstPaneItem in firstPane.getItems()
        if firstPaneItem.getPath() == item.getPath()
          pane.destroyItem item
        else
          pane.moveItemToPane item, firstPane
```

Sorry if this is a noob question, this is my first attempt at hacking something in atom

**Comment by lemongrassnginger** [comment]
hey @nwaywood, 

Have you considered putting the _close duplicate tabs_ functionality in a separate function? It would be useful as a command on it's own. You can then have another command which calls the merge panes functionality immediately followed by the close duplicates (that may have an undesirable flickering of the tabs though, not sure). If you don't beat me to it, I'll have a play with the code later

this isn't a noob question and in my opinion, noob questions are welcome here :) at least I hope they are...I ask them all the time!

**Comment by nwaywood** [comment]
@lemongrassnginger I have done some more work trying to get this working but still no luck...

- From some help over [here](https://discuss.atom.io/t/pane-destoryitem-not-working/48802) I found out that `destroyItem` returns a promise, so the core issue is that the command returns before the promise is resolved
- I tried using `await` keyword to wait for the promise to resolve. But then I found out that `await` was introduced in CoffeeScript 2 and Atom still is running CoffeeScript 1
- I then tried to get it working using generators in CoffeeScript 1 and failed horribly
- I then tried porting my init script to JS so I can use `async`/`await` and ended up with this code

```
atom.commands.add("atom-workspace", "custom:merge-panes", async () => {
  const panes = atom.workspace.getCenter().getPanes()
  const firstPane = panes.shift()

  // loop through all panes except for the first pane
  for (pane of panes) {
    for (item of pane.getItems()) {
      // if item is already in first pane, delete it, otherwise move it to first pane
      for (firstPaneItem of firstPane.getItems()) {
        if (firstPaneItem.getPath() === item.getPath()) {
          await pane.destroyItem(item)
        } else {
          pane.moveItemToPane(item, firstPane)
        }
      }
    }
  }
})
```

but that code doesn't work either :/

**Comment by nathansobo** [comment]
@nwaywood It looks like the problem is with the fact that you're moving the item in the inner loop when really you should wait until you don't find a duplicate and move it in the outer loop. I tweaked your logic a bit and have a command that seems to be working as you intended. I think I'll keep it in personal init script. If someone is interested in opening a PR with a command like this and CCing me on it, I will help get it merged into core.

```js
atom.commands.add("atom-workspace", "custom:merge-panes", async () => {
  const panes = atom.workspace.getCenter().getPanes()
  const firstPane = panes.shift()

  // loop through all panes except for the first pane
  for (const pane of panes) {
    for (const item of pane.getItems()) {
      // if item is already in first pane, delete it, otherwise move it to first pane
      if (firstPane.getItems().find((i) => i.getPath() === item.getPath())) {
        pane.destroyItem(item)
      } else {
        pane.moveItemToPane(item, firstPane)
      }
    }
  }
})
```

**Comment by nwaywood** [comment]
@nathansobo Yes that command works! Thank you!

I would be happy to submit a PR for it :) any chance you could point me in the right direction to which files I will need to modify?

**Comment by nwaywood** [comment]
fyi I have improved the logic a little so that it also works with non `TextEditor` items:

```js
atom.commands.add("atom-workspace", "custom:merge-panes", () => {
  const panes = atom.workspace.getCenter().getPanes()
  const firstPane = panes.shift()

  // loop through all panes except for the first pane
  for (pane of panes) {
    for (item of pane.getItems()) {
      // if item is already in first pane, delete it, otherwise move it to first pane
      if (firstPane.getItems().find(findItemInPane(item))) {
        pane.destroyItem(item)
      } else {
        pane.moveItemToPane(item, firstPane)
      }
    }
  }
})

function findItemInPane(nonFirstPaneItem) {
  return function(firstPaneItem) {
    // check that both items are standard editor items (e.g. not SettingsView)
    if (
      Object.getPrototypeOf(firstPaneItem).constructor.name === "TextEditor" &&
      Object.getPrototypeOf(nonFirstPaneItem).constructor.name === "TextEditor"
    ) {
      return firstPaneItem.getPath() === nonFirstPaneItem.getPath()
    } else {
      return firstPaneItem.uri === nonFirstPaneItem.uri
    }
  }
}
```

**Comment by ctf0** [comment]
- keymap
```coffee
'body':
  'cmd-k up': 'pane:split-up-and-move-active-item'
  'cmd-k down': 'pane:split-down-and-move-active-item'
  'cmd-k left': 'pane:split-left-and-move-active-item'
  'cmd-k right': 'pane:split-right-and-move-active-item'
  'cmd-k cmd-end': 'custom:pane-merge'
```

- init
```coffee
# add merge pane command
atom.commands.add 'atom-workspace',
    'custom:pane-merge', ->
      aPanes = atom.workspace.getCenter().getPanes()
      oFirstPane = aPanes.shift()
      for oPane in aPanes
        for oItem in oPane.getItems()
          oPane.moveItemToPane oItem, oFirstPane

# context menu
atom.contextMenu.add {
    'atom-text-editor, .overlay': [
        {label: 'Split Up',command: 'pane:split-up-and-move-active-item'},
        {label: 'Split Down',command: 'pane:split-down-and-move-active-item'},
        {label: 'Split Left',command: 'pane:split-left-and-move-active-item'},
        {label: 'Split Right',command: 'pane:split-right-and-move-active-item'},
        {label: 'Close Pane',command: 'custom:pane-merge'},
    ]
}
```
